Bug 10104 - Only restyle nodes that uses viewport percentage units on viewport size change

shinglyu opened the pull request:
<!-- Please describe your changes on the following line: -->
## Bug 10104 - Only restyle nodes that uses viewport percentage units on viewport size change

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #10104 (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/11890)

<!-- Reviewable:end -->

shinglyu committed:
Only restyle viewport-relative nodes on viewport size change

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties/longhand/background.mako.rs, components/style/properties/longhand/text.mako.rs, components/style/properties/longhand/position.mako.rs, components/style/properties/longhand/outline.mako.rs, components/style/properties/longhand/list.mako.rs, components/style/properties/properties.mako.rs, components/style/properties/data.py, components/style/traversal.rs, components/style/properties/longhand/font.mako.rs, components/style/properties/longhand/pointing.mako.rs, components/style/properties/longhand/inherited_box.mako.rs, components/style/properties/longhand/inherited_text.mako.rs, components/style/properties/longhand/effects.mako.rs, components/style/values.rs, components/style/properties/longhand/counters.mako.rs, components/style/dom.rs, components/style/properties/longhand/border.mako.rs, components/style/properties/longhand/box.mako.rs, components/style/properties/helpers.mako.rs, components/style/properties/longhand/inherited_table.mako.rs, components/style/properties/longhand/column.mako.rs
 * @kichjang: components/script/dom/node.rs, components/script/layout_wrapper.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

shinglyu commented:
r? @mbrubeck

shinglyu commented:
Profiling shows that the restyle events dropped from 

```
	_incremental?_	_iframe?_	_url_	_mean (ms)_	_median (ms)_	_min (ms)_	_max (ms)_	_events_
Compositing	     N/A	  N/A	             N/A	7.4163	0.1395	0.0003	53.598	97
Layout	 yes	yes	http://localhost:8000/mozilla.html	40.7831	0.1542	0.0282	1242.3715	76
Layout	 no	yes	http://localhost:8000/mozilla.html	2.7127	2.7127	2.7127	2.7127	1
+ Style Recalc	 yes	yes	http://localhost:8000/mozilla.html	14.7199	0.0465	0.0205	620.1162	97
+ Style Recalc	 no	yes	http://localhost:8000/mozilla.html	1.3714	1.3714	1.3714	1.3714	1
```

to

```
_category_	_incremental?_	_iframe?_	_url_	_mean (ms)_	_median (ms)_	_min (ms)_	_max (ms)_	_events_
Compositing	     N/A	  N/A	             N/A	4.3865	0.0998	0.0003	35.7685	33
Layout	 yes	yes	http://localhost:8000/mozilla.html	148.2419	0.1642	0.0456	1173.1	22
Layout	 no	yes	http://localhost:8000/mozilla.html	2.9211	2.9211	2.9211	2.9211	1
+ Style Recalc	 yes	yes	http://localhost:8000/mozilla.html	46.404	0.0288	0.0259	725.4902	29
+ Style Recalc	 no	yes	http://localhost:8000/mozilla.html	1.628	1.628	1.628	1.628	1
```

The HTML waterfall chart also showed improved restyle speed.

shinglyu commented on components/style/properties/properties.mako.rs:
```diff
@@ -822,6 +823,17 @@ impl ToCss for PropertyDeclaration {
+    fn has_viewport_percentage(&self) -> bool {
+        match *self {
+            % for property in data.longhands:
+                PropertyDeclaration::${property.camel_case}(DeclaredValue::Value(ref val)) => val.has_viewport_percentage(),
```
This line is too long, but wrapping this will make the generated code ugly. Suggestions welcome.

shinglyu commented:
Here are the profiler HTML file and tsv dump: 
[profiler_output.zip](https://github.com/servo/servo/files/336419/profiler_output.zip)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #11766) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #11972) made this pull request unmergeable. Please resolve the merge conflicts.

shinglyu commented:
ping @mbrubeck

mbrubeck commented:
I've started reviewing this and should have a complete review ready soon.

mbrubeck commented:
@bors-servo delegate+

This looks good! Can land with r=mbrubeck after rebasing and addressing the minor whitespace/comment issues below.

---

Reviewed 1 of 1 files at r1, 1 of 3 files at r7, 5 of 14 files at r9, 21 of 21 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[components/layout_thread/lib.rs, line 1073 \[r10\]](https://reviewable.io:443/reviews/servo/servo/11890#-KLxQp560l74AOLzK56x:-KLxQp560l74AOLzK56y:1158063403) ([raw file](https://github.com/servo/servo/blob/672dcf1afea4d5be31e4f8cb069e7cf81cc119e1/components/layout_thread/lib.rs#L1073)):*
> ```Rust
>                        .unwrap();
>             }
>             // FIXME (#10104): Only dirty nodes affected by vh/vw/vmin/vmax styles.
> ```

You can remove this comment.

---

*[components/layout_thread/lib.rs, line 1078 \[r10\]](https://reviewable.io:443/reviews/servo/servo/11890#-KMBaBmCFzY1GP4AW8RB:-KMBaBmCFzY1GP4AW8RC:827856191) ([raw file](https://github.com/servo/servo/blob/672dcf1afea4d5be31e4f8cb069e7cf81cc119e1/components/layout_thread/lib.rs#L1078)):*
> ```Rust
>                     if node.needs_dirty_on_viewport_size_changed() {
>                         node.dirty_self();
>                         node.dirty_descendants();
> ```

It would be nice if we could avoid traversing the node's descendants after they are already dirtied. Maybe this could be a follow-up enhancement.

---

*[components/style/properties/properties.mako.rs, line 830 \[r9\]](https://reviewable.io:443/reviews/servo/servo/11890#-KLKfP-XIf3-sivCHBF6-r9-830:-KMB_S2xlQkYZy2z8N17:-1928777853) ([raw file](https://github.com/servo/servo/blob/07426f85d4240185c3c0415559b66060d0f519b2/components/style/properties/properties.mako.rs#L830)):*
<details><summary><i>Previously, shinglyu (Shing Lyu) wrote…</i></summary>
> This line is too long, but wrapping this will make the generated code ugly. Suggestions welcome.
</details>
I'd just wrap it, and don't worry much about the generated code ugliness.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11890#-:-KMBaNwEUGwpPEiVHyQ2:1450329894)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:v: @shinglyu can now approve this pull request

shinglyu commented:
I'll land it first and make the skipping traversal part a followup, because this part of the code get bitrotted too fast.

shinglyu commented:
@bors-servo  r=mbrubeck

bors-servo commented:
:pushpin: Commit 720cf27 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 720cf27fadf4598b6497f3c5979115e887a2cb23 -->

bors-servo commented:
:hourglass: Testing commit 720cf27 with merge 901a6bd...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2025)

shinglyu commented:
Oops, forgot to check if geckolib builds or not.

shinglyu commented:
@mbrubeck I've updated the patch to support Stylo. One thing worth mentioning is that Gecko doesn't has the flag I just added, so I fallback to a full restyle when it is used in Stylo. Do you want to check again? Or I can just merge?

shinglyu commented:
@bors-servo r=mbrubeck

bors-servo commented:
:pushpin: Commit 5452bba has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 5452bbab7b16e30e4bcfcda1dfa56b01212e01f6 -->

bors-servo commented:
:hourglass: Testing commit 5452bba with merge d45080e...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/2058)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #12458) made this pull request unmergeable. Please resolve the merge conflicts.

shinglyu commented:
Implement the `HasViewportPercentage` trait for `vector_longhand` helper

shinglyu commented:
@bors-servo r=mbrubeck

bors-servo commented:
:pushpin: Commit f754cac has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck f754cacbd53c47c63639e1d1226812427ef1eaad -->

bors-servo commented:
:hourglass: Testing commit f754cac with merge d87ea67...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1935), [arm32](http://build.servo.org/builders/arm32/builds/2046), [arm64](http://build.servo.org/builders/arm64/builds/2048), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2090), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2210), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2119), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2107), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2178), [windows](http://build.servo.org/builders/windows/builds/1788)
